metricplugin/http.go

```diff
@@ -0,0 +1,326 @@
+
+// A decorator that ensures the server is not closing and, if not, wraps the
+// given function in a wait group.
+func (s *httpServer) ensureNotClosedAndWaitGroup(f apiFunction) apiFunction {
+	return func(r *http.Request, p httprouter.Params) (response, error) {
+		select {
+		case <-s.closing:
+			return nil, ErrServerClosing
+		default:
+		}
+
+		s.wg.Add(1)
+		defer s.wg.Done()
+
+		return f(r, p)
+	}
+}
+
+// A decorator that recovers from a panic and returns ErrInternalServerError.
+func recoverPanic(f apiFunction) apiFunction {
+	return func(r *http.Request, p httprouter.Params) (resp response, fErr error) {
+		defer func() {
+			err := recover()
+			if err != nil {
+				log.Errorf("recovered panic: %s", err)
+				resp = nil
+				fErr = ErrInternalServerError
+			}
+		}()
+
+		return f(r, p)
+	}
+}
+
+// A decorator that ensures errors are client-presentable.
+// Non-presentable errors are replaced with ErrInternalServerError.
+// After this, one of these cases applies:
+// - either response is nil and error is a presentableError, or
+// - response is not nil and error is nil.
+func ensurePresentableError(f apiFunction) apiFunction {
+	return func(r *http.Request, p httprouter.Params) (response, error) {
+		resp, err := f(r, p)
+		if err != nil {
+			if resp != nil {
+				panic("inner function returned an error and a non-nil response")
+			}
+
+			_, ok := err.(presentableError)
+			if !ok {
+				// Not a nice error. Replace with ISE.
+				log.Errorf("recovered non-presentable error: %s", err)
+				err = ErrInternalServerError
+			}
+		}
+
+		return resp, err
+	}
+}
+
+// A decorator that expands a (response, presentableError) tuple into a
+// (HTTP status code, response, error) triple.
+// If the error is not presentable, this panics.
+func extractStatusCode(f apiFunction) statusedResponseFunction {
+	return func(r *http.Request, p httprouter.Params) (int, response, error) {
+		status := http.StatusOK
+		resp, err := f(r, p)
+		if err != nil {
+			pErr, ok := err.(presentableError)
+			if !ok {
+				panic("API produced non-presentable error")
+			}
+			status = pErr.code()
+		}
+
+		return status, resp, err
+	}
+}
+
+// A decorator that encodes a generated response into a JSONResponse.
+func jsonEncodeResponse(f statusedResponseFunction) httprouter.Handle {
+	return func(w http.ResponseWriter, h *http.Request, p httprouter.Params) {
+		status, resp, err := f(h, p)
+
+		w.WriteHeader(status)
+
+		var errString *string
+		if err != nil {
+			s := err.Error()
+			errString = &s
+		}
+
+		err = json.NewEncoder(w).Encode(JSONResponse{
+			Status: status,
+			Result: resp,
+			Err:    errString,
+		})
+
+		if err != nil {
+			log.Errorf("unable to write JSON response to %s: %s", h.RemoteAddr, err)
+		}
+	}
+}
+
+type pingResponse struct{}
+
+func (pingResponse) response() {}
+
+type monitoringAddressesResponse struct {
+	Addresses []string `json:"addresses"`
+}
+
+func (monitoringAddressesResponse) response() {}
```

Comment:
Not really sure why this is empty...?

Reply:
It's a marker interface: I just want to explicitly mark things that are sent as responses. This serves as a type bound on some function signatures, instead of using `interface{}`, which I find annoying to reason about...